pkg/common/peertracker/peertracker_test_child_windows.go

```diff
@@ -0,0 +1,195 @@
+	flag.Parse()
+
+	if protocolInfoFile == "" {
+		fmt.Fprint(os.Stderr, "-protocolInfoFile flag is required")
+		os.Exit(1)
+	}
+
+	// We are a grandchild - send a sign then sleep forever
+	if tcpSocketPort == -1 {
+		piBytes, err := getProtocolInfo(protocolInfoFile)
+		if err != nil {
+			fmt.Fprintf(os.Stderr, "Error reading file with WSAPROTOCOL_INFO structure: %v", err)
+			os.Exit(2)
+		}
+
+		pi := (*windows.WSAProtocolInfo)(unsafe.Pointer(&piBytes[0]))
+		fd, err := windows.WSASocket(-1, -1, -1, pi, 0, 0)
+
+		if err != nil {
+			fmt.Fprintf(os.Stderr, "WSASocket error: %v", err)
+			os.Exit(3)
+		}
+
+		err = send(fd, "i'm alive!")
+		if err != nil {
+			fmt.Fprintf(os.Stderr, "Send failed: %v", err)
+			os.Exit(4)
+		}
+
+		// sleep forever
+		select {}
+	}
+
+	addr := &net.TCPAddr{
+		IP:   net.IPv4(127, 0, 0, 1),
+		Port: tcpSocketPort,
+	}
+
+	conn, err := net.DialTCP("tcp", nil, addr)
+	if err != nil {
+		fmt.Fprintf(os.Stderr, "DialTCP failed: %v", err)
+		os.Exit(5)
+	}
+
+	socket, err := getFd(conn)
+	if err != nil {
+		fmt.Fprintf(os.Stderr, "Failed to get file descriptor from connection: %v", err)
+		os.Exit(6)
+	}
+
+	procattr := &syscall.ProcAttr{
+		Env:   os.Environ(),
+		Files: []uintptr{0, 0, 0}, // Do not block on stdin / stdout / stderr
+	}
+
+	pid, _, err := syscall.StartProcess(os.Args[0], []string{os.Args[0], "-protocolInfoFile", protocolInfoFile}, procattr)
+	if err != nil {
+		fmt.Fprintf(os.Stderr, "Failed to produce grandchild: %v", err)
+		os.Exit(7)
+	}
+
+	if err := duplicateSocket(windows.Handle(socket), pid, protocolInfoFile); err != nil {
+		fmt.Fprintf(os.Stderr, "Failed to duplicate socket: %v", err)
+		os.Exit(8)
+	}
+
+	// Inform our caller of the grandchild pid
+	fmt.Fprintf(os.Stdout, "%v", pid)
+	os.Exit(0)
+}
+
+// wsaDuplicateSocket calls the WSADuplicateSocket function that returns
+// a WSAPROTOCOL_INFO structure that can be used to create a new socket
+// descriptor for a shared socket
+func wsaDuplicateSocket(socket windows.Handle, processID uint32, protocolInfo *windows.WSAProtocolInfo) (err error) {
+	r1, _, e1 := syscall.Syscall(procWSADuplicateSocketA.Addr(), 3, uintptr(socket), uintptr(processID), uintptr(unsafe.Pointer(protocolInfo)))
+	if r1 == 0xffffffff {
+		return e1
+	}
+	return nil
+}
+
+// getProtocolInfo gets the WSAPROTOCOL_INFO structure from
+// the file at protocolInfoFile path. If there is no content
+// after 15 seconds, this function times out
+func getProtocolInfo(protocolInfoFile string) (piBytes []byte, err error) {
+	timeout := time.After(15 * time.Second)
+	tick := time.Tick(500 * time.Millisecond)
+
+	for {
+		select {
+		case <-timeout:
+			return []byte{}, errors.New("timed out")
+		case <-tick:
+			piBytes, _ = os.ReadFile(protocolInfoFile)
+			if len(piBytes) > 0 {
+				return piBytes, nil
+			}
+		}
+	}
+}
+
+// getFd gets the file descriptor from a TCP connection
+func getFd(conn *net.TCPConn) (fd uintptr, err error) {
```

Comment:
This function is a little worrisome in that it returns the socket handle obtained via RawConn.Control, which breaks the guarantees the Control gives (namely that the handle is technically only guaranteed to remain valid for the duration of the Control call). Wonder if we souldn't do the dup'ing within the callback...

Reply:
Thanks for pointing this! Doing the duplication within the callback is the most convenient way.